triedb/pathdb/states.go

```diff
@@ -43,9 +49,528 @@ func (c *counter) report(count metrics.Meter, size metrics.Meter) {
+// an additional flag *destructed = true* will be returned, otherwise the flag is
+// false. Besides, the returned list will include the hash of deleted storage slot.
+// Note a special case is an account is deleted in a prior tx but is recreated in
+// the following tx with some storage slots set. In this case the returned list is
+// not empty but the flag is true.
+//
+// Note, the returned slice is not a copy, so do not modify it.
+//
+//nolint:unused
+func (s *stateSet) storageList(accountHash common.Hash) ([]common.Hash, bool) {
+	s.lock.RLock()
+	_, destructed := s.destructSet[accountHash]
+	if _, ok := s.storageData[accountHash]; !ok {
+		// Account not tracked by this layer
+		s.lock.RUnlock()
+		return nil, destructed
+	}
+	// If an old list already exists, return it
+	if list, exist := s.storageListSorted[accountHash]; exist {
+		s.lock.RUnlock()
+		return list, destructed // the cached list can't be nil
+	}
+	s.lock.RUnlock()
+
+	// No old sorted account list exists, generate a new one
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	storageList := maps.Keys(s.storageData[accountHash])
+	slices.SortFunc(storageList, common.Hash.Cmp)
+	s.storageListSorted[accountHash] = storageList
+	return storageList, destructed
+}
+
+// clearCache invalidates the cached account list and storage lists.
+func (s *stateSet) clearCache() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.accountListSorted = nil
+	s.storageListSorted = make(map[common.Hash][]common.Hash)
+}
+
+// merge integrates the accounts and storages from the external set into the
+// local set, ensuring the combined set reflects the combined state of both.
+//
+// The provided state set will remain unchanged, as it may still be referenced
+// by other layers.
+func (s *stateSet) merge(set *stateSet) {
+	var (
+		delta             int
+		accountOverwrites counter
+		storageOverwrites counter
+		destructs         []destruct
+	)
+	// Apply account deletion markers and discard any previously cached data if exists
+	for accountHash := range set.destructSet {
+		if origin, ok := s.accountData[accountHash]; ok {
+			delta -= common.HashLength + len(origin)
+			accountOverwrites.add(common.HashLength + len(origin))
+			delete(s.accountData, accountHash)
+		}
+		if _, ok := s.storageData[accountHash]; ok {
+			// Looping through the nested map may cause slight performance degradation.
+			// However, since account destruction is no longer possible after the cancun
+			// fork, this overhead is considered acceptable.
+			for _, val := range s.storageData[accountHash] {
+				delta -= 2*common.HashLength + len(val)
+				storageOverwrites.add(2*common.HashLength + len(val))
+			}
+			delete(s.storageData, accountHash)
+		}
+		// Keep track of whether the account has already been marked as destructed.
+		// This additional marker is useful for undoing the merge operation.
+		_, exist := s.destructSet[accountHash]
+		destructs = append(destructs, destruct{
+			Hash:  accountHash,
+			Exist: exist,
+		})
```

Comment:
I don't quite see why each change needs to be appended here. Don't they either all get reverted or none? 

The semantics of `merge` usually means that "we now much everything into one" -- so this need to keep a journal of events makes me confused.

Reply:
In `destructs`, it tracks the detail for each destruct operation, specifically whether the associated account has already been marked as destructed or not.

These information will be retained in the stateSet and will be useful when we do revert;

In Revert, we want to undo the destruction set updating and we must need this flag to 
distinguish:

(a) the account has already been marked as destructed before applying this merge operation
(`stateSet.destructSet[addr]` is existent) and undo merge won't clear this entry in the `stateSet.destructSet` map)

(b) the account hasn't been marked as destructed before applying this merge operation
(`stateSet.destructSet[addr]` is not existent) and undo merge should clear this entry in the `stateSet.destructSet` map)

Reply:
I still do not "see" it. Can you outline a scenario where this matters. I mean in terms like this: 

Block A: contract x is deleted. 
Block B: contract x is recreated
Block C: contract x is deleted
Block D: contract x is deleted

Now we merge `B` into `A`, stateset `AB`. And then merge `C` into `AB`, stateset `ABC`. Then revert `?`...

Reply:
Block A: contract x is deleted, and then recreated
Block B: contract x is deleted, and then recreated

We will have set A (destruct[x] = true, account[x] = x_1)
We will have set B (destruct[x] = true, account[x] = x_2)

We merge B into A, then AB is (destruct[x] = true, account[x] = x_2)

If we want to revert B from AB, then reverted set should be ((destruct[x] = true, account[x] = x_1)

Reply:
In another case, 

Block A: contract x is updated
Block B: contract x is deleted, and then recreated

We will have set A (destruct[x] = false, account[x] = x_1)
We will have set B (destruct[x] = true, account[x] = x_2)

We merge B into A, then AB is (destruct[x] = true, account[x] = x_2)

If we want to revert B from AB, then reverted set should be ((destruct[x] = **false**, account[x] = x_1)

Reply:
In the first case, 

> We merge B into A, then AB is (destruct[x] = true, account[x] = x_2)

When merging `B` into `A`, so that the "end state" represents all changes in A and B, shouldn't we wind up with `(destruct[x] = false, account[x] = x_2)` -- that is: the account is not destructed. The destruct+recreate simply becomes a modification. 

In the second case.  

> We merge B into A, then AB is (destruct[x] = true, account[x] = x_2)

In this instance, again, `update + delete + recreate` should IMO become `update`. 

That's how I think about the `merge` operation. Doing it this way means that the semantics are clear. And the same goes for slots, doesn't it? 

If, the following happens for slot `X`: `2->0`, `0->1`, it "might aswell" have gone `2->1`. We don't care about all intermediate transitions.

Reply:
> When merging B into A, so that the "end state" represents all changes in A and B, shouldn't we wind up with (destruct[x] = false, account[x] = x_2) -- that is: the account is not destructed. The destruct+recreate simply becomes a modification.

The AB should be `(destruct[x] = true, account[x] = x_2)`. `The destruct[x] = true` means the **original account x* ** was destructed in set AB and all the state access to x should be forbidden.

Reply:
For 1. the set should be `destruct[x] = true, account[x] = v`
For 2. the set should be `destruct[x] = false, account[x] = v`

Reply:
The `destruct[addr]` is only marked if this account is deleted within the block.
The `account[addr]` is only marked if this account is modified/created within the block.

These two flags can be combined

Reply:
>> This is what I don't understand. What is the difference between
>> 
>>   1. an  **original account `X`** that has been destructed and recreated, and 
>>   2. a modified account `X`.



> For 1. the set should be `destruct[x] = true, account[x] = v`
> For 2. the set should be `destruct[x] = false, account[x] = v`

That doesn't really explain the reason why. What I am wondering is this: why is it important whether something was, at some point, destructed, and then resurrected? 

As I see it, it is analoguous to the case in storage: `1 -> 0 -> 2` (via `0`), vs `1 -> 3 -> 2`. It doesn't matter whether one intermediate step was `0` or `3`, the important thing is that 

- it was modified, and end state is `2`.

So what I am asking is: what is the case where this makes a difference, and how would the code behave erroneously if we dropped this tracking?

Reply:
> What I am wondering is this: why is it important whether something was, at some point, destructed, and then resurrected?

Yes, it's important.

Let's say we have an account x, it's destructed at block y, and then recreated at block y; then the stateset of y should be `destruct[x] = true, account[x] = a2`

The destruction flag matters when the storage slots of x is accessed;

Let's say the original account x has three slots (s1, s2, s3); the recreated account x has two slots (s3, s4)

StorageSlot(x, s1, at block y) = null
StorageSlot(x, s2, at block y) = null
StorageSlot(x, s3, at block y) = new_slot_3
StorageSlot(x, s4, at block y) = new_slot_4

The destruction flag represents that **all the storage slots of the account has been deleted**, the slots in the deep layer belong to the original x, instead of the newly created x

Reply:
Btw, i think i can try to get rid of the destruct flag totally.

When an account is deleted, we can mark all its storage slots as deleted in the storageData set. Originally, we used a destruction flag to prevent excessive accumulation of items in memory, which could potentially lead to out-of-memory issues. However, with self-destruction deprecated, we now know that handling all storage deletions can fit comfortably within memory as it has in the past. No reason to have this additional flag for more complexity.

Reply:
I agree, that would simplify things. Do we know what the largest "delete-event" in mainnet history was, by any chance?

Reply:
Yes, https://gist.github.com/rjl493456442/5cf1d9a8783c33f500611e2b1663135e

Reply:
done